vshard/router/init.lua

```diff
@@ -873,36 +888,158 @@ local function router_map_callrw(router, func, args, opts)
                             router.total_bucket_count - bucket_count)
         goto fail
     end
+    do return timeout, nil, nil, rid, futures, replicasets end
+
+    ::fail::
+    return nil, err, err_id, rid, futures
+end
+
+--
+-- Perform Ref stage of the Ref-Map-Reduce process on a subset of all the
+-- replicasets, which contains all the listed bucket IDs.
+--
+local function router_ref_storage_by_buckets(router, bucket_ids, timeout)
+    local grouped_buckets
+    local group_count
+    local err, err_id, res
+    local replicasets_all = router.replicasets
+    local replicasets_to_map = {}
+    local futures = {}
+    local opts_async = {is_async = true}
+    local deadline = fiber_clock() + timeout
+    local rid = M.ref_id
+    M.ref_id = rid + 1
+
+    -- Nil checks are done explicitly here (== nil instead of 'not'), because
+    -- netbox requests return box.NULL instead of nils.
+
+    -- Ref stage.
+    while next(bucket_ids) do
+        -- Group the buckets by replicasets according to the router cache.
+        grouped_buckets, err = buckets_group(router, bucket_ids, timeout)
+        if grouped_buckets == nil then
+            goto fail
+        end
+        timeout = deadline - fiber_clock()
+
+        -- Netbox async requests work only with active connections.
+        -- So, first need to wait for the master connection explicitly.
+        local replicasets_to_check = {}
+        group_count = 0
+        for uuid, _ in pairs(grouped_buckets) do
+            group_count = group_count + 1
+            table.insert(replicasets_to_check, replicasets_all[uuid])
+        end
+        timeout, err, err_id = wait_connected_to_masters(
+            replicasets_to_check, timeout)
+        if not timeout then
+            goto fail
+        end
+
+        -- Send ref requests with timeouts to the replicasets.
+        futures = table_new(0, group_count)
+        for id, buckets in pairs(grouped_buckets) do
+            if timeout == nil then
+                err_id = id
+                goto fail
+            end
+            local args_ref
+            if replicasets_to_map[id] then
+                -- Replicaset is already referenced on a previous iteration.
+                -- Simply get the moved buckets without double referencing.
+                args_ref = {'moved_buckets', buckets}
```

Comment:
Good question, I didn't think of that or somewhy discarded the idea. Now I thought it through again, and to my surprise it was quite untrivial to find a solution.

In the result I kept 2 functions, but I replaced `moved_buckets` with another one, which not only checks moved buckets but also checks if the ref is still present. It was a non-critical bug, that if a storage ref has disappeared (timed out or the storage was restarted or anything), we would still check the moved buckets from the given list while even the previously confirmed buckets could already escape. The router could then even send Map requests and the ones going to the storages with no refs would guaranteed fail. Now the router can find it out earlier.

Then, I spent some time thinking where I could make it one function. Apparently, it would have to be doing this:
- Set ref if asked. But don't set otherwise. Because if a ref disappeared, we don't want to set it again. We want an error.
- Return moved buckets.
- Signal whether the ref was set (might be not set even if asked, if no suitable buckets were found here at all).

And for such a function I couldn't even find a proper name. `storage_ref_ensure_with_buckets()`? All variations of "ensure", "guarantee" wouldn't pass, because they mean "set ref no matter what" even if I only wanted to check its presence. Names containing "check" wouldn't pass either, because I sometimes do want to set a ref, not just check it.

Finally, I also realized that we would have to make it co-exist with the existing `storage_ref()` function. Or even merge them into one. Like make `bucket_ids` an option. But that would mean we need `storage_ref()` take its second argument either as a timeout-number or as a table of options, and then return either bucket count total or this weird map with moved buckets and whether a ref is done, and then my brain stopped.

In the end I decided to keep more than one function, but with simple tasks:
- `storage_ref()` works like before, unrelated to this MR.
- `storage_ref_make_with_buckets(rid, timeout, bucket_ids)` - it does just what it says.
- `storage_ref_check_with_buckets(rid, bucket_ids)` - the new one. Works like the older `moved_buckets()` but also checks the ref. Returns an error when the ref isn't found.

For that I also had to add one more commit which introduces `vshard.storage.ref.check()` + a couple of tests.

Reply:
Sounds reasonable, thank you for the detailed explanation